### PR TITLE
Fix for bug regarding externs types in generic procedures

### PR DESCRIPTION
**Pull request by bradcray**
Prior to this fix, using a generic extern type in a generic Chapel procedure resulted in
an internal error because extern generic types are treated as primitive types and
the compiler would try to clone the type for each generic instantiation, yet we don't
support cloning of primitive types.

This fix simply takes the approach of not creating a copy of an external type symbol
but returning the original symbol instead.

In addition to retiring the original future, I also added a new test that actually uses
the extern type to lock it in a bit further.

### CODE REVIEW

**Comment by bradcray** [comment]
I wondered a bit about this as well -- or rather, whether the logic to catch this case should occur further up in the callchain.  For the case in question, the callstack above this point is:

* symbol::copy, defined as a macro in symbol.h --- doing the check here felt morally equivalent to modifying copyInner() and had the additional obstacle of jamming it into the macro/header which felt yucky.  But I think even if I'd done so, you'd have the same hesitation since copy() and copyInner() are effectively two parts of the same logical operation.

* DefExpr::CopyInner() in expr.cpp -- here, we could introduce logic that said "if this DefExpr is defining a typeSymbol and that typeSymbol is an external  type then just use the original typeSymbol in the DefExpr rather than requesting a copy of the typesymbol and relying on the logic I inserted to return the original symbol.

As you mentioned over lunch, the current change has the downside of being potentially surprising, but also catching any possible call into the extern typesymbol copy would be caught.

Thinking about this some more, it also occurs to me that while we don't want to copy the typical primitive type, it might be OK to do so for external primitive types (and maybe this argues that external types shouldn't share representation with internal primitive types).  Why don't I try a different approach and see if we can copy such types by changing the logic that was causing the internal error and see if we like that better.

**Comment by bradcray** [comment]
@noakesmichael : See if you like this approach better.  I think I do.

**Comment by bradcray** [comment]
(currently testing)

File: compiler/AST/type.cpp [hunk 1]
```diff
@@ -117,8 +117,25 @@ PrimitiveType::PrimitiveType(Symbol *init, bool internalType) :
...
+  // suggest that external primitive types should really be
+  // represented as their own ExternType class...
+  //
+  if (!symbol->hasFlag(FLAG_EXTERN)) {
+    INT_FATAL(this, "Unexpected call to PrimitiveType::copyInner");
+  }
+
+  copy = new PrimitiveType(NULL);
+  return copy;
 }
```
Comment (noakesmichael) [inline compiler/AST/type.cpp]: Silly observation that doesn't need to be followed :- I would have skipped the local variable "copy" and written the
tail as

return new PrimitiveType(NULL);

**Comment by noakesmichael** [comment]
I agree that this seems sound.

File: compiler/AST/type.cpp (hunk 1 above)
Comment (bradcray) [inline compiler/AST/type.cpp]: Oh yeah, that is much better.  This is the result of saying "what do these other copyInner()s on types do, and then pattern matching.  Of course the other ones do nontrivial things with their 'copy' variable and then I didn't think about it hard enough after getting it working to note that I didn't (and didn't need to).

**Pull request merged**